### PR TITLE
[BugFix] Fix memory illegal modification

### PR DESCRIPTION
**Pull request by sevev**
When we create a primary key table, we will check the total length of the key columns and we don't support the length of the key column greater than 128 bytes. However, when the primary key table only has one key column and the column type is varchar, we may write some data whose key column length is larger than 128 bytes. So in some code of persistent index, we may create a vector that contains at most 129 elements and modify the vector by key column sizes

```
e.g.

// kMaxKeyLength is defined as 128
std::vector<std::pair<int64_t, int64_t>> add_usage_and_size(kMaxKeyLength + 1, std::pair<int64_t, int64_t>(0, 0));
....
add_usage_and_size[keys[i].size].first -= keys[i].size + kIndexValueSize;
add_usage_and_size[keys[i].size].second--;
```

So if we can not guarantee the size of the key column is not larger than 128 bytes, the above code is illegal memory modification which may cause unexpected errors.

## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.0
  - [x] 2.5
  - [x] 2.4
  - [x] 2.3

### CODE REVIEW

File: be/src/storage/primary_key_encoder.cpp [hunk 1]
```diff
@@ -526,9 +526,12 @@ bool PrimaryKeyEncoder::encode_exceed_limit(const Schema& schema, const Chunk& c
     std::vector<const void*> datas(ncol, nullptr);
     if (ncol == 1) {
         if (schema.field(0)->type()->type() == TYPE_VARCHAR) {
```
Comment (chaoyli) [inline be/src/storage/primary_key_encoder.cpp:528]: You can add a unit test case for it.

Comment (sevev) [reply be/src/storage/primary_key_encoder.cpp:528]: > You can add a unit test case for it.

done, finished in https://github.com/StarRocks/starrocks/pull/24680/commits/c138702863f4871bb32ef525a4ed8f2f9333d167

**Review by chaoyli — APPROVED** [review]

**Comment by decster** [comment]
> so if we can guarantee the size of the key column is not larger than 128 bytes, the above code is illegal memory modification which may cause unexpected errors.

If we cannot ?

**Comment by sevev** [comment]
> > so if we can guarantee the size of the key column is not larger than 128 bytes, the above code is illegal memory modification which may cause unexpected errors.
> 
> If we cannot ?

Yes, done.

**Review by decster — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.4

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.3

**Comment by wangsimo0** [comment]
branch 2.5: https://github.com/StarRocks/starrocks/pull/24791

**Pull request merged**